### DISCUSSION TITLE
Formula testing

### DISCUSSION THREAD

**Discussion by robertfiko** [Homebrew/discussions#1518] [Tap maintenance and Homebrew development]
I am writing a formula installer with dependecies:

```ruby
depends_on "yaws" => "2.0.9"
depends_on "erlang"
depends_on "gcc" => "4.9.1"
depends_on "graphviz" => "2.47.1"
```

Now I am in the phase of testing, and  it cannot start, as I get this error message:
```
Error: referl is missing test dependencies: erlang
```
(referl is te formula name)

Even for this little trivially bad test case:
```ruby
test do
   assert_true(false)
end
```

I am using this command for running test cases manually:
```
brew test referl.rb
```

What is problem there? Why it cannot find erlang for testing?

Thanks for any help!

**Comment by SMillerDev** [comment]
You haven't installed erlang. You also can't specify versions for dependencies so it'll probably complain about that after.

**Reply by robertfiko** [reply]
Thank you, but I have installed erlang on my machine, and also it is specified in the dependency list. 
Also, I have read in the Formula cookbook that it is possible to specify versions, so what is the problem with that, can you explain it please?

**Reply by SMillerDev** [reply]
> Thank you, but I have installed erlang on my machine, and also it is specified in the dependency list.

What happens when you `brew install` your formula?

https://docs.brew.sh/Formula-Cookbook#specifying-other-formulae-as-dependencies

The version specification here is only valid for symbols.

**Reply by robertfiko** [reply]
Thanks, for clearifying the version specificaiton.

I install it with `brew install --build-from-source referl.rb`, because I do not want to upload everytime to GitHub to test.
It says:

```
Updating Homebrew...
==> Auto-updated Homebrew!
Updated 1 tap (homebrew/core).
==> Updated Formulae
Updated 2 formulae.

Error: Failed to load cask: referl.rb
Cask 'referl' is unreadable: wrong constant name #<Class:0x00007f80828ed6b8>
Warning: Treating referl.rb as a formula.
Warning: referl 0.9.20.08 is already installed and up-to-date.
To reinstall 0.9.20.08, run:
 brew reinstall referl
```

**Reply by gromgit** [reply]
I thought you said it was a _formula_? Please post the full contents of `referl.rb`.

**Reply by carlocab** [reply]
> I install it with `brew install --build-from-source referl.rb`, because I do not want to upload everytime to GitHub to test.
> It says:

You don't need to make any changes to your GitHub repository to test. Just make the changes to your local repository containing the `referl` formula and do `brew [re]install referl`.

**Reply by robertfiko** [reply]
> I thought you said it was a _formula_? Please post the full contents of `referl.rb`.

I am writing a formula, am i not supposed to use `--build-from-source` to test the formula?

Any ways, here is the sourcecode
```ruby
# typed: false
# frozen_string_literal: false

# Refactoring software for Erlang language by ELTE and Ericson
class Referl < Formula
  desc "Refactoring software for Erlang language by ELTE and Ericson"
  homepage "https://plc.inf.elte.hu/erlang/index.html"
  url "https://plc.inf.elte.hu/erlang/dl/refactorerl-0.9.20.08_v2.zip"
  version "0.9.20.08"
  sha256 "52f0778c42f7c48490f93b07a435cb3f8c3573810765b6255145e6972edc0cea"
  license "LGPL-3.0-only" # SPDX Representation for: GNU Lesser General Public License v3.0 only

  depends_on "erlang"
  depends_on "gcc"
  depends_on "graphviz"
  depends_on "yaws"
  

  # Creating exec script
  def create_exec_script
    out_file = File.new("bin/#{runner_script}", "w")
    out_file.puts("#\!\/bin\/bash")
    out_file.puts("#{String.new(HOMEBREW_PREFIX)}/Cellar/referl/+#{version}/bin/referl_boot -base #{String.new(HOMEBREW_PREFIX)}/Cellar/referl/#{version}/ $@")
    out_file.close
  end

  def yaws_detect
    yaws_version = `yaws --version`.split[-1]
    yaws_path = "#{String.new(HOMEBREW_PREFIX)}/Cellar/yaws/#{yaws_version}/lib/yaws-#{yaws_version}/ebin"
    ohai "Looking for YAWS path on: #{yaws_path}"
    if File.directory?(yaws_path)
      ohai "Yaws found!"
    else
      odie("Error! - yaws path not found, you may need to install one manually")
    end
    yaws_path
  end

  def installpaths
    install_paths = []
    Dir["bin/*"].each { |x| install_paths.push(String.new(x)) }
    install_paths.delete("bin/referl")
    install_paths.delete("bin/#{runner_script}")
    install_paths
  end

  def instal_referl
    install_paths = installpaths

    # Installing
    bin.install install_paths
    bin.install "bin/referl" => "referl_boot" # Rename referl to avoid conflicts with symlink
    lib.install Dir["lib/*"]
    prefix.install "refactorerl.boot"
    prefix.install "sys.config"

    bin.install "bin/referl_exec" => "referl" # Symlink will be pointed to this script, due to rename
  end

  def install
    yaws_path = yaws_detect
    create_exec_script

    system "bin/referl", "-build", "tool", "-yaws_path", yaws_path
    instal_referl
  end

  test do
    assert_true(false)
    # Fobb esetek: elindul, args, -yaws, -db, -name, doksiban tobb
  end
end
```

**Reply by gromgit** [reply]
> I am writing a formula, am i not supposed to use `--build-from-source` to test the formula?

Yes, you'd have to do that to build the formula. I was referring to the following error that you posted:
```
Error: Failed to load cask: referl.rb
Cask 'referl' is unreadable: wrong constant name #<Class:0x00007f80828ed6b8>
```
I managed to replicate that error when referencing the Ruby file directly in a `brew install`, but putting your formula in the `Formula` directory of a tap (as @carlocab mentioned) causes it to `brew audit` and begin an install correctly, so it seems to point to an incorrect fallthrough in the Homebrew code.

**Reply by robertfiko** [reply]
Alright, I just wanted help with tests, but it seems, I am in a bigger trouble right now, but I cannot fuly understand. 
Where is the error? What should I look into or how should I start my journey to fix it?
(sorry if I missed something you mentioned, but I feel lost right now)

**Reply by gromgit** [reply]
Not an error _per se_, just a weird code path inside Homebrew that caused a cask-related complaint, even though `referl.rb` is clearly a formula. As @carlocab has already mentioned, Homebrew expects formulae to be hosted inside taps, rather than be standalone files in random locations. The simplest way around the above:
```
brew tap-new robertfiko/tap
mv referl.rb $(brew --repo robertfiko/tap)/Formula
brew install robertfiko/tap/referl
```
